Fix Node Event Handlers for Shift Click

Myestery opened the pull request:
This pull request refactors the node selection and pointer interaction logic in the Vue node graph editor to improve multi-selection behavior, clarify event handling, and enhance test coverage. The main change is to defer multi-select toggle actions (such as ctrl+click for selection/deselection) from pointer down to pointer up, preventing premature selection state changes and making drag interactions more robust. The drag initiation logic is also refined to only start dragging after the pointer moves beyond a threshold, and new composable methods are introduced for granular node selection control.

**Node selection and pointer event handling improvements:**
* Refactored multi-select (ctrl/cmd/shift+click) logic in `useNodeEventHandlersIndividual`: selection toggling is now deferred to pointer up, and pointer down only brings the node to front without changing selection state. The previous `hasMultipleNodesSelected` function and related logic were removed for clarity. [[1]](diffhunk://#diff-8d3820a1ca9c569bce00671fdd6290af81315ae11b8f3d6f29a5a9d30379d084L18-L35) [[2]](diffhunk://#diff-8d3820a1ca9c569bce00671fdd6290af81315ae11b8f3d6f29a5a9d30379d084L57-L73) [[3]](diffhunk://#diff-89bfc2a05201c6ff7116578efa45f96097594eb346f18446c70aa7125ab1811aL112-L116) [[4]](diffhunk://#diff-89bfc2a05201c6ff7116578efa45f96097594eb346f18446c70aa7125ab1811aR127-R143)
* Added new composable methods `deselectNode` and `toggleNodeSelectionAfterPointerUp` to `useNodeEventHandlersIndividual` for more granular control over node selection, and exposed them in the returned API. [[1]](diffhunk://#diff-8d3820a1ca9c569bce00671fdd6290af81315ae11b8f3d6f29a5a9d30379d084R210-R245) [[2]](diffhunk://#diff-8d3820a1ca9c569bce00671fdd6290af81315ae11b8f3d6f29a5a9d30379d084L251-R259)

**Pointer interaction and drag behavior changes:**
* Updated `useNodePointerInteractions` to track pointer down/up state and only start dragging after the pointer moves beyond a pixel threshold. Multi-select toggling is now handled on pointer up, not pointer down, and selection state is read from the actual node manager for accuracy. [[1]](diffhunk://#diff-b50f38fec4f988dcbee7b7adf2b3425ae1e40a7ff10439ecbcb380dfa0a05ee1R6-R10) [[2]](diffhunk://#diff-b50f38fec4f988dcbee7b7adf2b3425ae1e40a7ff10439ecbcb380dfa0a05ee1R33-R34) [[3]](diffhunk://#diff-b50f38fec4f988dcbee7b7adf2b3425ae1e40a7ff10439ecbcb380dfa0a05ee1R44-R53) [[4]](diffhunk://#diff-b50f38fec4f988dcbee7b7adf2b3425ae1e40a7ff10439ecbcb380dfa0a05ee1R76-R110) [[5]](diffhunk://#diff-b50f38fec4f988dcbee7b7adf2b3425ae1e40a7ff10439ecbcb380dfa0a05ee1R122-R123) [[6]](diffhunk://#diff-b50f38fec4f988dcbee7b7adf2b3425ae1e40a7ff10439ecbcb380dfa0a05ee1L131-R175)

**Test suite enhancements:**
* Improved and expanded tests for pointer interactions and selection logic, including new cases for ctrl+click selection toggling on pointer up, drag threshold behavior, and mocking of new composable methods. [[1]](diffhunk://#diff-8d94b444c448b346f5863e859c75f67267439a56a02baf44b385af1c6945effdR9-R11) [[2]](diffhunk://#diff-8d94b444c448b346f5863e859c75f67267439a56a02baf44b385af1c6945effdR35-R56) [[3]](diffhunk://#diff-8d94b444c448b346f5863e859c75f67267439a56a02baf44b385af1c6945effdR100-R102) [[4]](diffhunk://#diff-8d94b444c448b346f5863e859c75f67267439a56a02baf44b385af1c6945effdL144-R181) [[5]](diffhunk://#diff-8d94b444c448b346f5863e859c75f67267439a56a02baf44b385af1c6945effdL155-R196) [[6]](diffhunk://#diff-8d94b444c448b346f5863e859c75f67267439a56a02baf44b385af1c6945effdL196-R247) [[7]](diffhunk://#diff-8d94b444c448b346f5863e859c75f67267439a56a02baf44b385af1c6945effdL276-R336) [[8]](diffhunk://#diff-8d94b444c448b346f5863e859c75f67267439a56a02baf44b385af1c6945effdR348-R423)
* Updated test setup and assertions for node event handlers, ensuring selection changes are only triggered at the correct event phase and that drag and multi-select logic is covered. [[1]](diffhunk://#diff-89bfc2a05201c6ff7116578efa45f96097594eb346f18446c70aa7125ab1811aL4-R7) [[2]](diffhunk://#diff-89bfc2a05201c6ff7116578efa45f96097594eb346f18446c70aa7125ab1811aR92)

These changes make node selection more predictable and user-friendly, and ensure drag and multi-select actions behave consistently in both the UI and the test suite.

fix https://github.com/Comfy-Org/ComfyUI_frontend/issues/6128


https://github.com/user-attachments/assets/582804d0-1d21-4ba0-a161-6582fb379352

Myestery committed:
[refactor] Centralize node deselection logic and fix TypeScript issues

- Move node click deselection logic from useNodePointerInteractions to useNodeEventHandlers
- Fix TypeScript issues by properly accessing nodeData.value
- Remove unused isLGraphNode import and hasMultipleNodesSelected function
- Simplify node selection logic in handleNodeSelect
- Add handleNodeClickDeselect to centralize deselection handling

Myestery committed:
[refactor] Implement deferred shift-click selection toggle with drag threshold

- Defer shift-click selection toggle until pointer up to enable shift+drag
- Add 3px drag threshold to distinguish clicks from drags
- Track node selection state from LGraphNode (not Vue reactive data) for accuracy
- Add toggleNodeSelectionAfterPointerUp handler for deferred toggle logic
- Refactor handleNodeSelect to skip selection changes when shift is held

Behavior:
- Shift+click unselected node: Adds to selection (if not dragging)
- Shift+click selected node: Removes from selection (if not dragging)
- Shift+drag selected/unselected nodes: Drags without changing selection
- Regular click: Single select

Fixes shift-click toggle not working due to immediate selection changes
preventing drag detection, and fixes state sync issues between Vue
reactive data and LGraphNode state.

Myestery committed:
[test] Improve selection toggle tests with clear timing and behavior specs

- Rewrite test descriptions to specify pointer down vs pointer up behavior
- Replace negative assertions with positive behavior descriptions
- Add new test coverage for shift key and drag scenarios

Changes to useNodeEventHandlers.test.ts:
- "defers selection changes" → "on pointer down with ctrl+click: brings node to front only, no selection changes"
- Add shift key test for completeness
- "selects/deselects node" → "on pointer up with multi-select: selects/deselects node that was unselected/selected at pointer down"
- Clarify that regular clicks handled elsewhere (not by toggle handler)

Changes to useNodePointerInteractions.test.ts:
- Split single vague test into two specific scenarios
- Add "on ctrl+click: calls toggleNodeSelectionAfterPointerUp on pointer up (not pointer down)"
- Add "on ctrl+drag: does NOT call toggleNodeSelectionAfterPointerUp"

Test count: 14 → 20 tests (6 new/improved)
All tests now clearly specify WHEN (pointer down/up) and WHAT behavior occurs.

Myestery committed:
Merge remote-tracking branch 'origin/main' into toolbox-repositioning

Myestery committed:
[fix] Prevent drag state setup for pinned nodes

Fixed an issue where pinned nodes could be dragged despite being pinned.
The problem was that isPointerDown was set to true before checking if
the node was pinned, allowing handlePointerMove to trigger drag detection.

Changed the order of operations in handlePointerDown:
1. Track selection state
2. Call onNodeSelect (allows selecting pinned nodes)
3. Check if pinned and return early
4. Only then set up drag state (startPosition, isPointerDown, startDrag)

This ensures pinned nodes remain selectable but cannot be dragged.

Fixes browser_tests/tests/vueNodes/interactions/node/select.spec.ts:63

simula-r commented on src/renderer/extensions/vueNodes/composables/useNodePointerInteractions.test.ts:
```diff
@@ -141,9 +169,16 @@ describe('useNodePointerInteractions', () => {
     // Test pointer cancel - selection happens on pointer down
-    pointerHandlers.onPointerdown(createPointerEvent('pointerdown'))
+    pointerHandlers.onPointerdown(
+      createPointerEvent('pointerdown', { clientX: 100, clientY: 100 })
```
For these arbitrary xy positions, here and elsewhere in these tests, is there any way to use a locator or other? I ran into an issue recently where adding the vue nodes banner to the top offset the page, and broke a bunch of these types of tests.

If its just a literal pointer event that could be anywhere probably fine, but if its location speciifc (like on a node) think we should use a locator?

DrJKL commented on src/renderer/extensions/vueNodes/composables/useNodePointerInteractions.test.ts:
```diff
@@ -193,8 +234,17 @@ describe('useNodePointerInteractions', () => {
+    pointerHandlers.onPointerdown(
+      createPointerEvent('pointerdown', { clientX: 100, clientY: 100 })
+    )
+    await nextTick()
```
nit: Usually when I see `await nextTick()` in a test, it means that there's an action that could/should have used Vue test-utils (which does the awaiting for you)

Myestery committed:
feat: implement shift drag selection logic in useNodePointerInteractions

Myestery committed:
refactor: consolidate and harmonize node selection logic

- Extract isMultiSelectKey utility to eliminate duplication
- Move ensureNodeSelectedForShiftDrag to useNodeEventHandlers
- Standardize deselectNodes to use canvas.deselect() API
- Consolidate all selection decision-making in useNodeEventHandlers
- Update useNodePointerInteractions to delegate to selection handlers
- Update tests to match refactored implementation

No behavioral changes - all tests passing.

Myestery committed:
test: clear mocks for node selection before each test

Myestery committed:
Merge branch 'main' into toolbox-repositioning

Myestery committed:
fix: hide selection toolbox during drag operations

The toolbox was reappearing during shift-drag on single selections because
updateSelectionBounds() didn't check if dragging was in progress.

When ensureNodeSelectedForShiftDrag updates selection during drag start,
it triggers the selectionChanged watcher which calls updateSelectionBounds.
This was setting visible=true even though handleDragStart had just set it
to false.

Now updateSelectionBounds() checks isDragging before showing the toolbox,
preventing it from appearing during any drag operation.

Myestery committed:
feat: immediately add unselected node to multi-selection on shift-drag

Previously, shift-dragging an unselected node when a multi-selection existed
would not add the node to the selection until after drag threshold was exceeded.
This created a confusing UX where the node appeared to be dragged alone initially.

Now when shift-dragging an unselected node:
- If no selection exists: select just that node
- If single selection exists: add node to create multi-selection
- If multi-selection exists: immediately add node and drag all together

Also improved handleNodeSelect to select (not toggle) when multi-select key
is pressed, and simplified toggleNodeSelectionAfterPointerUp to only deselect
on pointer-up (since selection now happens on pointer-down).

Myestery committed:
fix: preserve multi-selection when dragging already-selected node

Myestery committed:
fix: handle single selection collapse to clicked node when multi-select is disabled
